### PR TITLE
go_install_from_commit: copy vendored dependencies into GOPATH

### LINKED ISSUE #64925: hack/verify-bazel.sh always fails on release-1.10
/kind bug
/sig testing
/sig release
/priority critical-urgent

```
I0608 15:49:50.952] +++ Running case: verify.bazel 
I0608 15:49:50.954] +++ working dir: /go/src/k8s.io/kubernetes
I0608 15:49:50.956] +++ command: bash "hack/make-rules/../../hack/verify-bazel.sh"
W0608 15:49:59.974] ../../internal/repos/remote.go:29:2: cannot find package "golang.org/x/tools/go/vcs" in any of:
W0608 15:49:59.975] 	/tmp/tmp.QZAPzQQRF1/go/src/github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/tools/go/vcs (vendor tree)
W0608 15:49:59.975] 	/usr/local/go/src/golang.org/x/tools/go/vcs (from $GOROOT)
W0608 15:49:59.975] 	/tmp/tmp.QZAPzQQRF1/go/src/golang.org/x/tools/go/vcs (from $GOPATH)
W0608 15:49:59.981] !!! [0608 15:49:59] Call tree:
W0608 15:49:59.983] !!! [0608 15:49:59]  1: ./hack/update-bazel.sh:34 kube::util::go_install_from_commit(...)
W0608 15:49:59.988] !!! [0608 15:49:59] Call tree:
W0608 15:49:59.990] !!! [0608 15:49:59]  1: ./hack/update-bazel.sh:34 kube::util::go_install_from_commit(...)
I0608 15:50:01.032] +++ exit code: 1
I0608 15:50:01.038] +++ error: 1
I0608 15:50:01.065] FAILED   verify-bazel.sh	11s
```

No PRs can be merged into release-1.10. See https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64848/pull-kubernetes-verify/94899/?log#log

Similar issue: https://github.com/kubernetes/kubernetes/issues/60730

/cc @ixdy

### PR DESCRIPTION
**Pull request by ixdy**
**What this PR does / why we need it**: `go_install_from_commit` is an abomination and I regret ever having written it.

gazelle recently vendored `golang.org/x/tools/go/vcs` (https://github.com/bazelbuild/bazel-gazelle/pull/228).

As a result, `go get -d github.com/bazelbuild/bazel-gazelle/cmd/gazelle` no longer installs `golang.org/x/tools/go/vcs` into the GOPATH, and so when we check out the older commit of `gazelle` to install, this dependency is missing.

This PR adds another level of hackery by copying any `vendor/` directory in the package path into the GOPATH before checking out the requested commit. This theoretically future-proofs us from similar breakages in the future, and using an arbitrary version of these dependencies from the `vendor/` directory is no worse than `go get -d` downloading the dependencies from tip. 

Cherry-picking #57600 is the better fix, but it's also more involved.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64925 

```release-note
NONE
```

/assign @cblecker @BenTheElder

### CODE REVIEW

File: hack/lib/util.sh [hunk 1]
```diff
@@ -474,9 +474,25 @@ kube::util::go_install_from_commit() {
...
   kube::util::ensure-temp-dir
   mkdir -p "${KUBE_TEMP}/go/src"
   GOPATH="${KUBE_TEMP}/go" go get -d -u "${pkg}"
+  # If the requested package has vendored dependencies, go get -d won't install
+  # them into GOPATH. This may break things if the dependencies weren't vendored
+  # at the commit requested.
+  # As a hacky workaround, copy the dependencies from tip into GOPATH.
+  (
+    cd "${KUBE_TEMP}/go/src/${pkg}"
+    while [[ "$(pwd)" != "${KUBE_TEMP}/go/src" && "$(pwd)" != "/" ]]; do
```
Comment (ixdy) [inline hack/lib/util.sh]: I know temp directories on macOS are weird, so I'd appreciate someone with a Mac testing to make sure this works correctly.

Comment (BenTheElder) [reply hack/lib/util.sh]: will do. also @cblecker

**Comment by ixdy** [comment]
/hold

The cherrypick PR https://github.com/kubernetes/kubernetes/pull/64929 is probably a better fix.

File: hack/lib/util.sh (hunk 1 above)
Comment (BenTheElder) [reply hack/lib/util.sh]: held to test https://github.com/kubernetes/kubernetes/pull/64929#event-1671529512 instead, running a few things to sanity check on mac

**Comment by ixdy** [comment]
/close

in favor of #64929

**Pull request closed without merging**